### PR TITLE
[FIX] delivery: make field visible to remove magic and make configurable

### PR DESCRIPTION
**Pull request by wtaferner**
In case a delivery method was not applied on an order before
confirmation the field invoice_shipping_on_delivery was set to True
which implied an extra line for each delivery fulfilled. This could mean
that an order which was partially fulfilled was getting multiple
shipping lines or an order which was maybe not intended to be billed for
shipping was getting automatically shipping costs applied.

Both situation are rational, but the user needs to have the choice to
either remove the automatically set option after confirmation or does
apply this option on purpose even before confirmation.

Adding the field in the view is an easy and correct fix as it makes the
magic somehow transparent and editable.

@mart-e @nim-odoo 
May I ask you to merge this for me? This one was producing a LOT of troubles. The solution is simpler than I expected and will finally make the feature complete. It is a typical UI Bug. Another solution would be to remove the code/feature, but I think this would be worse in terms of stable policy.

Info: @wt-io-it

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by wtaferner** [comment]
@nhomar @pedrobaeza @Yenthe666 
You might be interested in this one...

**Comment by Yenthe666** [comment]
I don't think it does too much harm and it can make things more clear. For me it looks ok.

**Comment by rim-odoo** [comment]
> In case a delivery method was not applied on an order before
> confirmation the field invoice_shipping_on_delivery was set to True
> which implied an extra line for each delivery fulfilled. This could mean
> that an order which was partially fulfilled was getting multiple
> shipping lines or an order which was maybe not intended to be billed for
> shipping was getting automatically shipping costs applied.

Well, actually, it is the intended behaviour. 
Either you set delivery costs at the quotation time.
Either you invoice each picking when it is sent.

> Both situation are rational, but the user needs to have the choice to
> either remove the automatically set option after confirmation or does
> apply this option on purpose even before confirmation.

This is more a functional change than a technical. So I would say "no" in a stable version. But that's super easy to customize if really needed (this PR is the proof ;) ).

File: addons/delivery/views/delivery_view.xml [hunk 1]
```diff
@@ -275,6 +275,7 @@
                 <data>
                   <xpath expr="//field[@name='payment_term_id']" position="after">
                       <field name="id" invisible="1"/>
+                      <field name="invoice_shipping_on_delivery" string="Extra line on each delivery" attrs="{'readonly': [('state', 'in', ('done'))]}"/>
```
Comment (rim-odoo) [inline addons/delivery/views/delivery_view.xml]: Your `string` is imprecise. I would rephrase: "Delivery orders will add shipping costs"

Comment (wtaferner) [reply addons/delivery/views/delivery_view.xml]: Hmm, it is still not precise and clear IMHO if I rephrase like you suggest.

What about: "Each delivery will add extra shipping costs"

**Comment by wtaferner** [comment]
@nim-odoo 
Hmm, could be true although one can argue if you lock a SO, it can also mean that nobody is supposed to change characteristics except invoicing stuff (shipping costs are part of sales)

@rim-odoo 

> Well, actually, it is the intended behaviour.
> Either you set delivery costs at the quotation time.
> Either you invoice each picking when it is sent.

Yes, it is for sure intended, but it is missing the clear user communication and information. 
How should someone know that this mechanism is in place, except finding it out at some point like Sherlock Holmes or simply in the code or wondering why the heck this always happens unwanted and finally do an extra module to display it.
So, no functional change but an UI bug and completion of an incomplete UI implementation ;-) 

Believe me, it was pure magic (from a user PoV and foremost not editable or changeable if someone took a wrong turn), even for me and I play around with Odoo quite a while in very different situations.

It is up to you guys where you merge it, but I claim it is a bug :stuck_out_tongue:

**Comment by wtaferner** [comment]
@rim-odoo @nim-odoo 
Do we get stuck here or is it merged anywhere...only a merged PR is a good PR in the end.

**Comment by rim-odoo** [comment]
My previous answer is still valid:

> This is more a functional change than a technical. So I would say "no" in a stable version.

**Comment by wtaferner** [comment]
@rim-odoo 
Ok, so, here we go ;-) Are you going to handle it?

**Comment by wtaferner** [comment]
Closing b/c it won't be merged in master neighter. Let's see if I need to save some users from this magic in v12 as well ;-)
